d3-shape/index.d.ts

```diff
@@ -1,380 +1,2321 @@
+     */
+    padAngle(angle: (this: This, data: Datum[], ...args: any[]) => number): this;
 }
 
+/**
+ * Constructs a new pie generator with the default settings.
+ *
+ * Ensure that the accessors used with the pie generator correspond to the arguments passed into them,
+ * or set them to constants as appropriate.
+ */
 export function pie(): Pie<any, number | { valueOf(): number }>;
+/**
+ * Constructs a new pie generator with the default settings.
+ *
+ * Ensure that the accessors used with the pie generator correspond to the arguments passed into them,
+ * or set them to constants as appropriate.
+ *
+ * The generic refers to the data type of an element in the input array passed into the Pie generator.
+ */
 export function pie<Datum>(): Pie<any, Datum>;
+/**
+ * Constructs a new pie generator with the default settings.
+ *
+ * Ensure that the accessors used with the pie generator correspond to the arguments passed into them,
+ * or set them to constants as appropriate.
+ *
+ * The first generic corresponds to the type of the "this" context within which the pie generator and its accessor functions will be invoked.
+ *
+ * The second generic refers to the data type of an element in the input array passed into the Pie generator.
+ */
 export function pie<This, Datum>(): Pie<This, Datum>;
 
 // -----------------------------------------------------------------------------------
 // Line Generators
 // -----------------------------------------------------------------------------------
 
-
+/**
+ * The line generator produces a spline or polyline, as in a line chart.
+ * Lines also appear in many other visualization types, such as the links in hierarchical edge bundling.
+ *
+ * The generic refers to the data type of an element in the input array passed into the line generator.
+ */
 export interface Line<Datum> {
-    (data: Array<Datum>): string | undefined;
-    x(): (d: Datum, index: number, data: Array<Datum>) => number;
+    /**
+     * Generates a line for the given array of data. Depending on this line generator’s associated curve,
+     * the given input data may need to be sorted by x-value before being passed to the line generator.
+     *
+     * IMPORTANT: If the rendering context of the line generator is null,
+     * then the line is returned as a path data string.
+     *
+     * @param data Array of data elements.
+     */
+    (data: Datum[]): string | null;
+    /**
+     * Generates a line for the given array of data. Depending on this line generator’s associated curve,
+     * the given input data may need to be sorted by x-value before being passed to the line generator.
+     *
+     * IMPORTANT: If the line generator has been configured with a rendering context,
+     * then the line is rendered to this context as a sequence of path method calls and this function returns void.
+     *
+     * @param data Array of data elements.
+     */
+    (data: Datum[]): void;
```

Comment:
You found the little slide of hands :wink:. Again, as with the above comment this is a more general pattern I adopted for those generators, which can render to `canvas` or create an `svg` path string.

I played around with it before settling on this approach as follows:
* You are right that technically, the signature for the `canvas` context which returns `void` is not reached by the compiler. At least in VS Code, it still shows up as an existing overload. Also, using the generator in the way that is idiomatic for `canvas` does not create a compile time error, even though the compiler assumes the first signature implicitly, i.e. the assumed return value of `string | null` goes wasted, as it is not assigned to anything.
* The first signature works with the common use case of passing the generator into a path element attribute setter, e.g. continuing the `arc` example from above:
```ts
pArc.attr('d', svgArc); // works!
````
* When I started the `strictNullChecks` validation, the generators were indeed returning `string | undefined` from when I initially wrote them as one-liner signatures. The problem with `undefined` becomes that `attr(...)` does not currently allow `undefined` as a type feeding its second setter argument. I did not want to add it, so the idiomatic usage above, would not work anymore. The issue is compounded by the fact that the post-fix operator `!` does _not_ work, i.e.:
```ts
pArc.attr('d', svgArc!); // does not compile!
````
That's due to the generators dual role as both object and function.

As an FYI the reason why I used `string | null` rather that just `string` is that, the d3 code for all the affected generators actually allows for a failure to generate the svg path string using its internal rendering buffer, in which case it return `null`.